invokeai/app/invocations/latent.py

```diff
@@ -695,27 +731,11 @@ def prep_ip_adapter_data(
             return None
 
         ip_adapter_data_list = []
+        assert len(ip_adapter) == len(image_prompts)
         for single_ip_adapter in ip_adapter:
-            ip_adapter_model: Union[IPAdapter, IPAdapterPlus] = exit_stack.enter_context(
-                context.models.load(single_ip_adapter.ip_adapter_model)
-            )
+            ip_adapter_model = exit_stack.enter_context(context.models.load(single_ip_adapter.ip_adapter_model))
 
-            image_encoder_model_info = context.models.load(single_ip_adapter.image_encoder_model)
-            # `single_ip_adapter.image` could be a list or a single ImageField. Normalize to a list here.
-            single_ipa_image_fields = single_ip_adapter.image
-            if not isinstance(single_ipa_image_fields, list):
-                single_ipa_image_fields = [single_ipa_image_fields]
-
-            single_ipa_images = [context.images.get_pil(image.image_name) for image in single_ipa_image_fields]
-
-            # TODO(ryand): With some effort, the step of running the CLIP Vision encoder could be done before any other
-            # models are needed in memory. This would help to reduce peak memory utilization in low-memory environments.
-            with image_encoder_model_info as image_encoder_model:
-                assert isinstance(image_encoder_model, CLIPVisionModelWithProjection)
-                # Get image embeddings from CLIP and ImageProjModel.
-                image_prompt_embeds, uncond_image_prompt_embeds = ip_adapter_model.get_image_embeds(
-                    single_ipa_images, image_encoder_model
-                )
+            image_prompt_embeds, uncond_image_prompt_embeds = image_prompts.pop(0)
```

Comment:
Are we popping to save memory? I'm guessing not, since the effect would probably be tiny relative to everything else in memory at this time. If not, it's generally better for functions not to have side effects of modifying their arguments (unless that is their explicit purpose).

Reply:
This is exactly the purpose -- I'm taking the image prompts off the stack one by one.

Reply:
I looked at it again and realized I could use zip() in the for loop, so I removed the pop.